### PR TITLE
Feat/gather scatter

### PR DESCRIPTION
**Pull request by nathanielsimard**
Make indexing functions more similar to other frameworks.

### CODE REVIEW

File: burn-tensor/src/tensor/api/check.rs [hunk 1]
```diff
@@ -374,6 +374,98 @@ impl TensorCheck {
         check
     }
 
+    pub fn gather<const D: usize>(dim: usize, shape: &Shape<D>, shape_indexes: &Shape<D>) -> Self {
```
Comment (antimora) [inline burn-tensor/src/tensor/api/check.rs:377]: I think we should start putting some minimal doc strings for all public APIs. Good pattern to establish.

File: burn-tensor/src/tensor/api/check.rs [hunk 2]
```diff
@@ -374,6 +374,98 @@ impl TensorCheck {
...
+        Self::check_gather_scatter_indexes(Self::Ok, "Gather", dim, shape, shape_indexes)
+    }
+
+    pub fn scatter<const D: usize>(
```
Comment (antimora) [inline burn-tensor/src/tensor/api/check.rs:381]: same

File: burn-tensor/src/tensor/api/check.rs [hunk 3]
```diff
@@ -374,6 +374,98 @@ impl TensorCheck {
...
+                    "The shape differs: {:?} != {:?}",
+                    shape_indexes.dims, shape_value.dims
+                )),
+            );
+        }
+
+        check
+    }
+
+    pub fn index_select<const D: usize>(dim: usize) -> Self {
```
Comment (antimora) [inline burn-tensor/src/tensor/api/check.rs:408]: same as above

File: burn-tensor/src/tensor/api/check.rs [hunk 4]
```diff
@@ -374,6 +374,98 @@ impl TensorCheck {
...
+        Self::check_index_select_basic::<D>(Self::Ok, "index_select", dim)
+    }
+
+    pub fn index_select_assign<const D: usize>(dim: usize) -> Self {
```
Comment (antimora) [inline burn-tensor/src/tensor/api/check.rs:412]: Doc

**Review by antimora — CHANGES_REQUESTED** [review]
Looks good. Requesting to add missing doc strings for the public API.

File: burn-tensor/src/tensor/api/check.rs (hunk 1 above)
Comment (nathanielsimard) [reply burn-tensor/src/tensor/api/check.rs:377]: I don't think this should be a public API, just pub(crate) I'll do the fix

**Review by antimora — APPROVED** [review]

**Pull request merged**